### PR TITLE
HdfsOpenWrite implementation similar to read

### PR DESCRIPTION
**Pull request by skibaa**
I used the same idea which works for HdfsOpenRead to implement the write operation.

### CODE REVIEW

**Comment by tmylk** [comment]
Looks promising. 
May I ask you to [update](https://github.com/RaRe-Technologies/smart_open/blob/master/smart_open/tests/test_smart_open.py#L458) existing tests and add new ones like for [read with mocking]
(https://github.com/RaRe-Technologies/smart_open/blob/master/smart_open/tests/test_smart_open.py#L141)

**Comment by skibaa** [comment]
Done

File: smart_open/tests/test_smart_open.py [hunk 1]
```diff
@@ -486,6 +485,19 @@ def test_s3_mode_mock(self, mock_write, mock_boto):
         mock_boto.connect_s3().get_bucket.assert_called_with("mybucket")
         self.assertTrue(mock_write.called)
 
+    @mock.patch('smart_open.smart_open_lib.subprocess')
+    def test_hdfs(self, mock_subprocess):
+        """Is HDFS write called correctly"""
+        smart_open_object = smart_open.HdfsOpenWrite(smart_open.ParseUri("hdfs:///tmp/test.txt"))
```
Comment (tmylk) [inline smart_open/tests/test_smart_open.py]: Woud it be possible to check on ` smart_open.smart_open("hdfs:///some/file.txt", "wb")`?

File: smart_open/smart_open_lib.py [hunk 2]
```diff
@@ -490,6 +492,33 @@ def __exit__(self, type, value, traceback):
         pass
 
 
+class HdfsOpenWrite(object):
+    """
+    Implement streamed reader from HDFS, as an iterable & context manager.
```
Comment (tmylk) [inline smart_open/smart_open_lib.py]: You mean `Implement streamed writer to HDFS, as a context manager.`?

File: smart_open/tests/test_smart_open.py (hunk 1 above)
Comment (skibaa) [inline smart_open/tests/test_smart_open.py]: I took example from the HdfsOpenRead, and asked myself why it was called directly and not via smart_open.smart_open. And answered to myself that in the unit test it is preferable to check the most direct API - the smart_open.smart_open, the call to ParseUri and branching according to the schema is tested elsewhere.

Comment (tmylk) [inline smart_open/tests/test_smart_open.py]: Yes, agree that HdfsOpenRead doesn't have this test. There is benefit in adding a test like that though to replace the exception test. As an aside, an ideal would be to mock hdfs similarly to [s3 tests]( https://github.com/skibaa/smart_open/blob/0fe9e521b5f7cf3589f6c5b2c537acec875a765d/smart_open/tests/test_smart_open.py#L251)

Comment (skibaa) [inline smart_open/tests/test_smart_open.py]: I did not understand - what is the outcome. Do you want to leave it as is or to replace the call to HdfsOpenWrite ctor to the smart_open call?

Comment (tmylk) [inline smart_open/tests/test_smart_open.py]: I would like to add tests for `smart_open.smart_open("hdfs:///some/file.txt", "wb")` and `smart_open.smart_open("hdfs:///some/file.txt", "rb")`

Comment (skibaa) [inline smart_open/tests/test_smart_open.py]: It is a good idea to add these tests. But IMO it's better to leave this pull request focus on the write operation alone, as it was from the beginning and add new tests for smart_open.smart_open on a separate pull request.

**Comment by tmylk** [comment]
Ok, Thanks for the PR!

**Pull request merged**